### PR TITLE
Fixes #3678 Add Generator Metatag

### LINKED ISSUE #3678: Add Generator Metatag
## Motivation
In quickstart 1, we had a metatag that indicated the generator of the markup:
```
<meta name="generator" content="UA Quickstart 7 (https://quickstart.arizona.edu)" />
```
We have had requests for a similar generator tag for Quickstart 2 to determine sites that are using the quickstart distribution.

### Is your feature request related to a problem? Please describe.
It would be useful for web teams on campus to be able to identify if a site is a Quickstart site.

## Proposed Resolution
Add a generator metatag to Quickstart 2.

### Describe the solution you'd like
Notably, we should not add any information about version or patch level, as this would be an unnecessary information disclosure.

### Describe alternatives you've considered
Not having a generator tag, with the view that this is information disclosure.

### PR DESCRIPTION
**Pull request by tadean**
## Description
This PR alters the Generator metatag from the system module to report the distribution name.

### Release notes
This change should probably have release notes. TBD.

## Related issues
Closes #3678 

## How to test

- Build site
- Inspect the html source
- Verify the presence of a generator tag with the following markup:
```
<meta name="Generator" content="Arizona Quickstart (https://quickstart.arizona.edu)" />
```

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] My change requires release notes.

### CODE REVIEW

**Review by bberndt-uaz — APPROVED** [review]

**Review by joeparsons — DISMISSED** [review]

**Review by trackleft — DISMISSED** [review]

**Comment by tadean** [comment]
- [ ] Test with `remove_http_headers`
- [ ] Test with `remove_generator`
- [ ] Possibly add config setting to turn off functionality?

@joshuasosa brought up that some environments will be using modules to turn off functionality like this. We should check to ensure they continue to work.

**Comment by accesswatch** [comment]
I approve. I have verified this result.

**Comment by tadean** [comment]
Added configuration setting to make generator tag optional.

**Comment by accesswatch** [comment]
Can I understand why this is optional by chance please? I am not challenging but just trying to understand t he harm it may cause by always having it?


From: tadean ***@***.***>
Sent: Tuesday, September 3, 2024 10:03 PM
To: az-digital/az_quickstart ***@***.***>
Cc: Bishop, Jeff - (jeffbis) ***@***.***>; State change ***@***.***>
Subject: [EXT] Re: [az-digital/az_quickstart] Fixes #3678 Add Generator Metatag (PR #3680)


External Email

________________________________
Added configuration setting to make generator tag optional.

-

**Comment by tadean** [comment]
> Can I understand why this is optional by chance please? I am not challenging but just trying to understand t he harm it may cause by always having it?

The rationale is that for public facing web applications, it's sometimes considered desirable to reveal as little information about the backend infrastructure of the site as possible, because knowledge of what framework a web application uses can inform attackers about potential vulnerabilities in the application. Some site owners may not want to disclose that their site is using Quickstart.

We received feedback from @joshuasosa about this particular issue that in their use case they take steps to hide the existing Drupal generator tag for this reason. 

Currently, the default for the optional configuration setting I've added is to be enabled, so it would be on for default unless the site owner takes steps to disable it. I can revert the change making it optional if it's not desirable.

**Review by trackleft — APPROVED** [review]

File: modules/custom/az_core/az_core.module [hunk 1]
```diff
@@ -372,6 +372,34 @@ function _az_core_check_attribute_form(array &$form, FormStateInterface $form_st
...
+      if (is_array($head_item) && in_array('system_meta_generator', $head_item, TRUE)) {
+        // Update the tag if it's present.
+        if (!empty($head_item[0]['#attributes']['content'])) {
+          $head_item[0]['#attributes']['content'] = 'Arizona Quickstart (https://quickstart.arizona.edu)';
+        }
+      }
+    }
+  }
+}
+
```
Comment (trackleft) [inline modules/custom/az_core/az_core.module:402]: Do you think we should add something like `$page['#cache']['tags'][] = 'config:az_core.settings';`

Comment (tadean) [reply modules/custom/az_core/az_core.module:402]: Good idea @trackleft !

**Comment by tadean** [comment]
- Removed optional configuration setting following code review
- Replaced the X-Generator header via service decoration
- Verified that the `remove_http_headers` module continues to work correctly with this functionality

**Review by bberndt-uaz — APPROVED** [review]

**Review by trackleft — APPROVED** [review]

**Pull request merged**